### PR TITLE
Add fontconfig

### PR DESCRIPTION
**Pull request by jakirkham**
Adds a recipe to build font-config. This is loosely based off of one in conda-recipes and includes some from my own recipe. Has been cleaned up so that it will fix nicely here at conda-forge.

### CODE REVIEW

**Comment by jakirkham** [comment]
@ccordoba12, would it be alright if I add you as a maintainer on this recipe?

Also, was trying to get a little more history on these patches. One is clearly custom for conda-build so won't be submitted upstream. However, the other one must have come from somewhere, but am having trouble figuring out where it came from and why it is needed. Do you know anything about this?

**Comment by jakirkham** [comment]
Looks like this is the [same patch]( https://sources.gentoo.org/cgi-bin/viewvc.cgi/gentoo-x86/media-libs/fontconfig/files/fontconfig-2.11.1-fc-cache-r.patch?view=markup ). Basically this is some sort of fix that went in after the release to solve [this bug]( https://bugs.freedesktop.org/show_bug.cgi?id=77252 ). Will write this up in the recipe. @ccordoba12, if you could confirm that my understanding of this is correct or otherwise clarify the nature of this patch, that would be great.

**Comment by ccordoba12** [comment]
@jakirkham, I really don't remember, but most probably I added it because I found it when looking at the Gentoo sources :-)

File: recipes/fontconfig/build.sh [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+
+
+if [[ `uname` == 'Darwin' ]];
+then
+    export LIBRARY_SEARCH_VAR=DYLD_FALLBACK_LIBRARY_PATH
+elif [[ `uname` == 'Linux' ]];
+then
+    export LIBRARY_SEARCH_VAR=LD_LIBRARY_PATH
+fi
```
Comment (ccordoba12) [inline recipes/fontconfig/build.sh]: Why is this needed?

File: recipes/fontconfig/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    - test -f "${PREFIX}/lib/libfontconfig.so"     # [linux]
+
+about:
+  home: http://www.freedesktop.org/wiki/Software/fontconfig/
+  license: MIT
+  summary: A library for configuring and customizing font access.
+
+extra:
+  recipe-maintainers:
+    - jakirkham
```
Comment (ccordoba12) [inline recipes/fontconfig/meta.yaml]: Please add me as a maintainer :-)

File: recipes/fontconfig/build.sh (hunk 1 above)
Comment (jakirkham) [inline recipes/fontconfig/build.sh]: So, it turns out if something tries to load the libraries we have built before fixing them with `install_name_tool` or `patchelf`, these may not work properly in some cases. While I am still a bit fuzzy on details, I believe this has to do with some dependency libraries using rpaths. By doing this, we can work around this issue during testing. It ends up being needed for install in this case as some binaries are run during the install process that run into issue otherwise. Fortunately, this has no effect on the package. After running `install_name_tool` or `patchelf` on the libraries and binaries following the build step, we find this is no longer necessary in the final package.

File: recipes/fontconfig/meta.yaml (hunk 2 above)
Comment (jakirkham) [inline recipes/fontconfig/meta.yaml]: Absolutely, thanks, @ccordoba12.

**Comment by jakirkham** [comment]
> I really don't remember, but most probably I added it because I found it when looking at the Gentoo sources :-)

That's fine. It has been a long time. Just trying to get this documented so we can understand what is going on with these patches and if/when these patches can be dropped.

File: recipes/fontconfig/meta.yaml (hunk 2 above)
Comment (jakirkham) [inline recipes/fontconfig/meta.yaml]: Added. :smile:

**Comment by pelson** [comment]
> That's fine. It has been a long time. Just trying to get this documented so we can understand what is going on with these patches and if/when these patches can be dropped.

Are they easy to test? i.e. if we remove the patch prematurely, can we make the build fail?

**Comment by jakirkham** [comment]
So, the Linux patch `fc-cache.patch` should make Linux fail if removed. This is literally from a commit that came after the release though. In other words already upstream and already cited in a bug it fixed on Linux.

The `fcf.patch` is trying to handle the fact that `fonts.conf` will be in a different location if we install this in a new prefix. So, we need to be able to change the binary to handle this. Hence that patch is used to insert the full prefix at build time, identify and handle it at packaging time, and properly set it at install time.

This all being said, I am unclear on whether we should apply them on Mac as well. The `fcf.patch` seems useful in all cases. As for the `fc-cache.patch`, this was to solve a Linux bug explicitly though maybe it is useful on Mac, as well. Any thoughts on this, @ccordoba12?

File: recipes/fontconfig/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  url: http://www.freedesktop.org/software/fontconfig/release/fontconfig-{{ version }}.tar.bz2
+  md5: 824d000eb737af6e16c826dd3b2d6c90
+  patches:
+    # Fixes a bug seen on Linux ( https://bugs.freedesktop.org/show_bug.cgi?id=77252 ).
+    # The patch is already present in upstream. See link below.
+    # ( https://cgit.freedesktop.org/fontconfig/commit/?id=f44157c809d280e2a0ce87fb078fc4b278d24a67 ).
+    - fc-cache.patch  # [linux]
+    # This patch won't be submitted upstream. It's use here is to ensure the prefix used
+    # to refer to the `fonts.conf` file.
+    - fcf.patch       # [linux]
```
Comment (jakirkham) [inline recipes/fontconfig/meta.yaml]: @ccordoba12, do you have any idea if these should be needed on Mac? I would think so, but am not sure.

File: recipes/fontconfig/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+build:
+  number: 0
+  skip: true                      # [win]
+  binary_has_prefix_files:        # [linux]
+    - lib/libfontconfig.so.1.8.0  # [linux]
```
Comment (jakirkham) [inline recipes/fontconfig/meta.yaml]: Why was this needed, @ccordoba12?

**Comment by jakirkham** [comment]
Could you please help us shed some light on the `fc-cache.patch`, @ccordoba12? We are trying to understand why this was necessary and what it fixed.

**Comment by jakirkham** [comment]
So, I don't know if we are going to hear back soon on this, @pelson. Not to pressure, but this is holding up way more interesting this like GraphViz. Before we hold this up much longer, I would like to understand something first. Is your objection a strong don't merge one or could it go in an issue on the feedstock?

File: recipes/fontconfig/fcf.patch [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+--- src/fccfg.c
++++ src/fccfg.c
+@@ -1830,9 +1830,7 @@ DllMain (HINSTANCE hinstDLL,
+ 
+ #endif /* !_WIN32 */
+ 
+-#ifndef FONTCONFIG_FILE
+-#define FONTCONFIG_FILE	"fonts.conf"
+-#endif
++#define FONTCONFIG_FILE	"@PREFIX@/etc/fonts/fonts.conf"
```
Comment (jakirkham) [inline recipes/fontconfig/fcf.patch]: @ccordoba12, could you please explain this line to me?

**Comment by jakirkham** [comment]
Since I'm not hearing much about this of late (last comment other than my own is over 2 wks old). I'm going to assume this is ok. Please add issues to the feedstock if you still have concerns. Thanks everyone for your feedback.

File: recipes/fontconfig/fcf.patch (hunk 5 above)
Comment (ccordoba12) [inline recipes/fontconfig/fcf.patch]: This is needed to avoid an error shown in the terminal when other programs load fontconfig dynamic library.
